src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -581,20 +581,23 @@ trait Namers extends MethodSynthesis {
         }
       }
 
-      def noDuplicates(names: List[Name], check: DuplicatesErrorKinds.Value) {
-        def loop(xs: List[Name]): Unit = xs match {
+      def noDuplicatesBy(namer: ImportSelector => Name, check: DuplicatesErrorKinds.Value): Unit = {
+        def canSkip(name: Name, sels: List[ImportSelector]): Boolean =
+          name == null || name == nme.WILDCARD || sels.forall(name ne namer(_))
+        @tailrec
+        def loop(xs: List[ImportSelector]): Unit = xs match {
           case Nil      => ()
           case hd :: tl =>
-            if (hd == nme.WILDCARD || !(tl contains hd)) loop(tl)
-            else DuplicatesError(tree, hd, check)
+            if (canSkip(namer(hd), tl)) loop(tl)
```

Comment:
Since we're traversing `tl` in `canSkip`, isn't this quadratic in time? (Not that it wasn't already.)

That probably doesn't matter, though, because (I'm on a phone and not that familiar with this code) this is usually called with a small list, right? (AFAIU `selectors` are the bracketed selectors in an import declaration.) Since my experience is that the common cases are that of a single non-renamed import and that of a wildcard import, if this is such a hot spot it may be beneficial to see if we can avoid getting into this logic at all.

Reply:
It already was, and still,  quadratic in the number of elements, in the worst case. Indeed, the worst case is what happens with normal scala code. How many elements, may depend on the codebase and the programmer: personally I do not usually write lists with over five imports, but some may very well prefer long lists to blanket `._` imports. 

The changes simply remove the auxiliary lists. The price of removing the auxiliary list is that there can be up to N calls to either `_.name` or `_.rename`, for each element. However, these are just class fields, not computations. 

A nice way to reduce the time complexity to an often linear could be to build a manual  "bloom filter", such as a Long `var`iable, and perform a run through the list that would set one bit of the filter: if there are no collisions, that would indicate no duplicates, and it would work in linear time.

Reply:
Yes, I'm a fan of the `noDuplicatesBy(_.rename)` for aesthetic reasons, too. I can't remember seeing more than 10 or so, and I found that unusual. (The fact that scalafmt then insists that each selector deserves its very own line makes the bracket syntax more verbose than repeating the import prefix all Java-like, ironically. (That is ironic, right? I remember being told by a particularly particular teacher of English that Alanis was mistaken about irony, but ironically I never remember why.))

Anyhow, I admit to having stepped in the amateur's death-trap of not considering constant factors, which probably trump any sort of asymptotics here. That's what profiling is for.